### PR TITLE
use weak pointer for callbacks

### PR DESCRIPTION
**Pull request by mrkraimer**
Use weak pointers for callbacks.

### CODE REVIEW

**Comment by mdavidsaver** [comment]
The title grabbed my attention as I have a perhaps similar problem with pva2pva which led me to https://github.com/mdavidsaver/pva2pva/blob/master/p2pApp/weakset.h

However, the code changes don't seem terribly focused.  So I would just caution you to remember the rules for iterator validity if unlocking in a loop, or allowing callbacks to remove themselves.  And also to cleanup dead weak pointers in long lived lists.

If you want more in the way of comment than this you'll need to provide some more details.

**Comment by mrkraimer** [comment]
On 01/12/2016 01:28 PM, mdavidsaver wrote:
>
> The title grabbed my attention as I have a perhaps similar problem 
> with pva2pva which led me to 
> https://github.com/mdavidsaver/pva2pva/blob/master/p2pApp/weakset.h
>
> However, the code changes don't seem terribly focused. So I would just 
> caution you to remember the rules for iterator validity if unlocking 
> in a loop, or allowing callbacks to remove themselves. And also to 
> cleanup dead weak pointers in long lived lists.
>
> If you want more in the way of comment than this you'll need to 
> provide some more details.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/epics-base/pvDatabaseCPP/pull/4#issuecomment-171003390>.
>

I have code like:
typedef std::tr1::weak_ptr<PVRecordClient> PVRecordClientWPtr;
...

         std::list<PVRecordClientWPtr>::iterator clientIter;
         while(true) {
             clientIter = pvRecordClientList.begin();
             if(clientIter==pvRecordClientList.end()) break;
             PVRecordClientPtr client = clientIter->lock();
             if(client.get()) {
                 unlock();
                 client->detach(getPtrSelf());
                 lock();
             }
             pvRecordClientList.erase(clientIter);
         }

Is this OK?
As far as I know I have no dead weak pointers in long lived lists.

Marty

**Comment by mdavidsaver** [comment]
> Is this OK?

Where I have gotten myself into trouble is iterating with callbacks, where the callback can modify the container.  Then looking at just the loop doesn't give enough information.  Since I see *unlock()* what about concurrent access?  And what is *client->detach()* allowed to change?

One of the ways to avoid this is to use a temporary list.  Since the temporary is a local there is no worry about unexpected concurrent access, and you don't even need to lock to fetch the next element.

> list<X> temp;
> temp.swap(pvRecordClientList);
> // maybe do something to prevent new clients from being added
> // unlock and iterate temp

For reference, a decent summary of iterator invalidation rules http://kera.name/articles/2011/06/iterator-invalidation-rules/ and for answering specific questions there are several good STL reference sites.  The one I usually look at first has a section 'Iterator validity' for each container method.  For example list::erase http://www.cplusplus.com/reference/list/list/erase/

Some other comments.

I know I've beaten you up about this before, but you really need to think about exception safety.  Having explicit _unlock()_ and _lock()_ looks like an exception thrown by *client->detach()* wouldn't re-lock.  So the code which eventually catches this exception might assume it has a lock which it doesn't.  This will led to a messy, hard to troubleshoot, crash.

The simple way to avoid this whole category of potential bugs is to use "guard" objects, like epicsGuard<> and epicsGuardRelease<> for mutexes.  When there isn't an existing existing class, then write one.  This is typically less than 10 lines of code.  Please start using this pattern.

For the example you give using iterators is unnecessary, and makes the code larger (source and object).  Better just to use _empty()_, _front()_, and _pop_front()_.

> while(!container.empty()) {
>   PVRecordClientPtr client = container.front().lock();
>   container.pop_front(); // since we have a local ref., can drop the weak_ptr immediately
>   // .. do stuff with client
> }

**Comment by mrkraimer** [comment]
On 01/13/2016 10:25 AM, mdavidsaver wrote:
>
>     Is this OK?
>
> Where I have gotten myself into trouble is iterating with callbacks, 
> where the callback can modify the container. Then looking at just the 
> loop doesn't give enough information. Since I see /unlock()/ what 
> about concurrent access? And what is /client->detach()/ allowed to change?
>
> One of the ways to avoid this is to use a temporary list. Since the 
> temporary is a local there is no worry about unexpected concurrent 
> access, and you don't even need to lock to fetch the next element.
>
>     list temp;
>     temp.swap(pvRecordClientList);
>     // maybe do something to prevent new clients from being added
>     // unlock and iterate temp
>
> For reference, a decent summary of iterator invalidation rules 
> http://kera.name/articles/2011/06/iterator-invalidation-rules/ and for 
> answering specific questions there are several good STL reference 
> sites. The one I usually look at first has a section 'Iterator 
> validity' for each container method. For example list::erase 
> http://www.cplusplus.com/reference/list/list/erase/
>
> Some other comments.
>
> I know I've beaten you up about this before, but you really need to 
> think about exception safety. Having explicit /unlock()/ and /lock()/ 
> looks like an exception thrown by /client->detach()/ wouldn't re-lock. 
> So the code which eventually catches this exception might assume it 
> has a lock which it doesn't. This will led to a messy, hard to 
> troubleshoot, crash.
>

My goal was to ensure thread and exception safety.
As far as I know I succeeded.


> The simple way to avoid this whole category of potential bugs is to 
> use "guard" objects, like epicsGuard<> and epicsGuardRelease<> for 
> mutexes. When there isn't an existing existing class, then write one. 
> This is typically less than 10 lines of code. Please start using this 
> pattern.
>
> For the example you give using iterators is unnecessary, and makes the 
> code larger (source and object). Better just to use /empty()/, 
> /front()/, and /pop_front()/.
>
>     while(!container.empty()) {
>     PVRecordClientPtr client = container.front().lock();
>     container.pop_front(); // since we have a local ref., can drop the
>     weak_ptr immediately
>     // .. do stuff with client
>     }
>

There may be a better way to implement thread and exception safety.
If someone can show a better way then great.

But remember the following requirements:

1) clients can connect and disconnect from a PVRecord.
2) Records can dynamically be added and removed from a PVDatabase.
When a record is removed all clients connected to the record must be 
notified.
3) Only one client at  a time can get data from, put data to, and 
process a record.
4) A client can connect to an arbitrary subset of fields in a record.
5) For gets and monitors only the data for fields that change value are 
returned to the client

Besides pvDatabaseCPP the code in pvDataCPP/src/copy is involved with 
implementing the requirements

File: src/database/pvRecord.cpp [hunk 1]
```diff
@@ -77,22 +77,28 @@ void PVRecord::destroy()
...
         while(true) {
             clientIter = pvRecordClientList.begin();
             if(clientIter==pvRecordClientList.end()) break;
+            PVRecordClientPtr client = clientIter->lock();
+            if(client.get()) {
+                unlock();
+                client->detach(getPtrSelf());
+                lock();
+            }
             pvRecordClientList.erase(clientIter);
```
Comment (mdavidsaver) [inline src/database/pvRecord.cpp]: The most obvious issue here is that explicit unlock()/lock() leaves open the possibility that an exception thrown be 'client->detach()' would no re-lock.  There are several ways to correct this.  One is explicit try/catch/rethrow.

```c++
if(client) {
  unlock();
  try {
    client->detach(getPtrSelf());
    lock();
  } catch(...) {
    lock();
    throw;
  }
}
```
I personally find this verbose and potentially error prone as the re-lock() call should be duplicated.  My preference is to add a "guard" class to handle unlock() and re-lock().  So a definition is added to some header.

```c++
template<typename T>
struct XUnLocker {
  T* ref;
  XUnLocker(T& r) :ref(&r) { r.unlock(); }
 ~XUnLocker() { r.lock(); }
};
```
Then the code in question becomes:

```c++
if(client) {
  XUnLocker<PVRecord> L(*this); // calls unlock()
  client->detach(getPtrSelf());
  // lock() when 'L' goes out of scope due to normal flow or exception unwind
}
```

Comment (mrkraimer) [inline src/database/pvRecord.cpp]: On 01/19/2016 11:47 AM, mdavidsaver wrote:
>
> In src/database/pvRecord.cpp 
> <https://github.com/epics-base/pvDatabaseCPP/pull/4#discussion_r50138258>:
>
> >          while(true) {
> >              clientIter = pvRecordClientList.begin();
> >              if(clientIter==pvRecordClientList.end()) break;
> > +            PVRecordClientPtr client = clientIter->lock();
> > +            if(client.get()) {
> > +                unlock();
> > +                client->detach(getPtrSelf());
> > +                lock();
> > +            }
> >              pvRecordClientList.erase(clientIter);

This appears in

> void PVRecord::destroy()
> {
>     if(traceLevel>1) {
>         cout << "PVRecord::destroy() " << recordName << endl;
>     }
>     lock();
>     try {
>         if(isDestroyed) {
>             unlock();
>             return;
>         }
>         isDestroyed = true;
>         pvTimeStamp.detach();
>
>         std::list<PVRecordClientWPtr>::iterator clientIter;
>         while(true) {
>             clientIter = pvRecordClientList.begin();
>             if(clientIter==pvRecordClientList.end()) break;
>             PVRecordClientPtr client = clientIter->lock();
>             if(client.get()) {
>                 unlock();
>                 client->detach(getPtrSelf());
>                 lock();
>             }
>             pvRecordClientList.erase(clientIter);
>         }
>         std::list<PVListenerWPtr>::iterator listenerIter;
>         while(true) {
>             listenerIter = pvListenerList.begin();
>             if(listenerIter==pvListenerList.end()) break;
>             PVListenerPtr listener = listenerIter->lock();
>             if(listener.get()) {
>                 unlock();
>                 listener->unlisten(getPtrSelf());
>                 lock();
>             }
>             pvListenerList.erase(listenerIter);
>
>         }
>         pvRecordStructure->destroy();
>         pvRecordStructure.reset();
>         pvStructure.reset();
>         unlock();
>     } catch(...) {
>         unlock();
>         throw;
>     }
> }

So the code IS enclosed in
try {
    ...
} catch(...) {
         unlock();
         throw;
     }

Note also that  the way lock is implemented it is OK to call unlock 
twice after a call to lock.
See pvDataCPP/src/misc/lock.h

**Comment by anjohnson** [comment]
On the subject of guard classes, Base already provides epicsGuard.h which implements both `template <class T> class epicsGuard;` and `template <class T> class epicsGuardRelease;`. Since this well-tested code is present in all current Base branches I would encourage its re-use instead of creating new implementations.

**Comment by ralphlange** [comment]
+1

**Comment by mrkraimer** [comment]
At the weekly hangout (2016.01.19) I did agree that I would back out the changes for weak pointers for callbacks.

I am ready to do this but want to make one more plea to allow this change.

If the concerns Michael raised are true than the concerns are valid with or without the weak pointers.

The weak pointers do offer one advantage,
If the code that implements the callback has gone away no callback is called.
If a shared rather than weak pointer is used than the callbacks are always called.
This assumes that the code that implements the callback must be implemented so that it is OK to call the callback.

So far this is OK because pvAccessCPP is prepared for this.
If the future this may not be true.

So one more time.

Can I  merge this change?

**Comment by anjohnson** [comment]
Hi Marty,

You are the module owner for pvDatabase, so you actually have the final say; you don't have to ask our permission to make changes to your own modules.

Looking at the actual changes I can only see one place where you're adding related calls to `lock();` and `unlock();` wrapped in/around a `try {} catch {}` block. However there are many places in your new version where you add calls to `iter->lock();` or construct a `shared_pointer` with `requester = somethingRequester.lock();` for which I don't see any equivalent calls to an `unlock()` method. That code may well be correct but it does seem strange (code smell), can you explain what is going on there? For example in src/pvAccess/channelLocal.cpp:

``` c++
void ChannelGetLocal::get()
{
    ChannelGetRequester::shared_pointer requester = channelGetRequester.lock();
    if(isDestroyed) {
```

If you can explain those instances I will be satisfied that this merge is OK.

However I still agree with Michael that using guard objects would make the code safer and easier to read. I have gone through pvDatabaseCPP and found many places where an epicsGuard and/or an epicsGuardRelease object can be used instead of a `try { ... } catch { unlock(); }` block. The diffstat for my changes ignoring fixes to indentation look like this:

```
tux% git diff -b | diffstat
 database/pvDatabase.cpp     |   51 ++++------------------------
 database/pvRecord.cpp       |   79 ++++++++------------------------------------
 pvAccess/channelLocal.cpp   |   74 ++++++++++-------------------------------
 pvAccess/monitorFactory.cpp |    8 +---
 4 files changed, 44 insertions(+), 168 deletions(-)
```

It looks like you once created a guard class in pvDataCPP; The header pv/lock.h defines a `class Lock` which appears to be very similar to the epicsGuard template and is used in many places, although it may not be quite as flexible as Jeff's template class. It would be nice to remove this kinds of duplication where it isn't strictly necessary.

\- Andrew

**Comment by mrkraimer** [comment]
On 01/19/2016 05:24 PM, Andrew Johnson wrote:
>
> Hi Marty,
>
> You are the module owner for pvDatabase, so you actually have the 
> final say; you don't have to ask our permission to make changes to 
> your own modules.
>
> Looking at the actual changes I can only see one place where you're 
> adding related calls to |lock();| and |unlock();| wrapped in/around a 
> |try {} catch {}| block. However there are many places in your new 
> version where you add calls to |iter->lock();| or construct a 
> |shared_pointer| with |requester = somethingRequester.lock();| for 
> which I don't see any equivalent calls to an |unlock()| method. That 
> code may well be correct but it does seem strange (code smell), can 
> you explain what is going on there? For example in 
> src/pvAccess/channelLocal.cpp:
>
> void  ChannelGetLocal::get()
> {
>      ChannelGetRequester::shared_pointer requester = channelGetRequester.lock();
>      if(isDestroyed) {
>
> If you can explain those instances I will be satisfied that this merge 
> is OK.
>

Note that what  what channelGetRequester.lock() does if to create a 
shared_ptr from a weak_ptr.
Thus lock has a different meaning that what we are used to with other 
epics code.

The code
ChannelGetRequester::shared_pointer requester = channelGetRequester.lock();
creates requester, i. e. the shared_ptr, on the stack
When the stack is cleared either because the method  completes or an 
exception is thrown requester is automatically destroyed.

Please respond if this satisfies your concern.


> However I still agree with Michael that using guard objects would make 
> the code safer and easier to read. I have gone through pvDatabaseCPP 
> and found many places where an epicsGuard and/or an epicsGuardRelease 
> object can be used instead of a |try { ... } catch { unlock(); }| block.
>

Again let me give requirements
1) clients can connect and disconnect from a PVRecord. 2) Records can 
dynamically be added and removed from a PVDatabase. When a record is 
removed all clients connected to the record must be notified. 3) Only 
one client at a time can get data from, put data to, and process a 
record.4) A client can connect to an arbitrary subset of fields in a 
record. 5) For gets and monitors only the data for fields that change 
value are returned to the client Besides pvDatabaseCPP the code in 
pvDataCPP/src/copy is involved with implementing the requirement 3) and 
4) and 5) I worked really hard to meet all of these requirement. There 
may be a better way but someone more knowledgeable then me is required.

**Comment by anjohnson** [comment]
Ok, so in this case the `lock()` method name comes from std::tr1::weak_ptr, and for as long as the `shared_ptr requester` variable exists on the stack it's temporarily incrementing the reference count of the referenced channelGetRequester object. The requester is thus acting as a guard object and provides exception-safety for this aspect of the implementation.

On the subject of replacing various `try { ... } catch { unlock(); }` blocks in pvDatabaseCPP with epicsGuard objects, I'll post a separate pull request with my suggested changes for you to look at after you have merged this branch. I won't be changing anything to do with what gets locked or when, just how the locks get applied and removed again to make the exception safety implicit instead of explicit and to reduce the amount of code necessary.

Ok, I'm happy with this. @mdavidsaver do you have any remaining objections to merging this weak_ptr stuff?

**Comment by mrkraimer** [comment]
I am going to merge and close this pull request.
I want this done because I am almost ready to push other changes to the master branch on mrkraimer/pvDatabaseCPP:
1) example are removed because they now are in exampleCPP
2) test is changed to me a true regression test.

I think that the weak pointer changes should be done before all the other changes.
